prime-router/src/main/kotlin/azure/WorkflowEngine.kt

```diff
@@ -288,6 +304,161 @@ class WorkflowEngine(
+                        context,
+                        options,
+                        actionHistory,
+                        txn
+                    )
+                }
+        }
+    }
+
+    // 1. create <event, report> pair or pairs depending on input
+    // 2. dispatchReport
+    // 3. log
+    private fun sendToDestination(
+        report: Report,
+        receiver: Receiver,
+        context: ExecutionContext,
+        options: Options,
+        actionHistory: ActionHistory,
+        txn: DataAccessTransaction
+    ) {
+        val loggerMsg: String
+        when {
+            options == Options.SkipSend -> {
+                // Note that SkipSend should really be called SkipBothTimingAndSend  ;)
+                val event = ReportEvent(Event.EventAction.NONE, report.id)
+                this.dispatchReport(event, report, actionHistory, receiver, txn, context)
+                loggerMsg = "Queue: ${event.toQueueMessage()}"
+            }
+            receiver.timing != null && options != Options.SendImmediately -> {
+                val time = receiver.timing.nextTime()
+                // Always force a batched report to be saved in our INTERNAL format
+                val batchReport = report.copy(bodyFormat = Report.Format.INTERNAL)
+                val event = ReceiverEvent(Event.EventAction.BATCH, receiver.fullName, time)
+                this.dispatchReport(event, batchReport, actionHistory, receiver, txn, context)
+                loggerMsg = "Queue: ${event.toQueueMessage()}"
+            }
+            receiver.format == Report.Format.HL7 -> {
+                report
+                    .split()
+                    .forEach {
+                        val event = ReportEvent(Event.EventAction.SEND, it.id)
+                        this.dispatchReport(event, it, actionHistory, receiver, txn, context)
+                    }
+                loggerMsg = "Queued to send immediately: HL7 split into ${report.itemCount} individual reports"
+            }
+            else -> {
+                val event = ReportEvent(Event.EventAction.SEND, report.id)
+                this.dispatchReport(event, report, actionHistory, receiver, txn, context)
+                loggerMsg = "Queued to send immediately: ${event.toQueueMessage()}"
+            }
+        }
+        context.logger.info(loggerMsg)
+    }
+
+    /**
+     * Handle a receiver specific event. Fetch all pending tasks for the specified receiver and nextAction
+     *
+     * @param messageEvent that was received
+     * @param context execution context
+     * @param actionHistory action history being passed through for this message process
+     */
+
+    fun handleProcessEvent(
+        messageEvent: ProcessEvent,
+        context: ExecutionContext,
+        actionHistory: ActionHistory
+    ) {
+        val errors: MutableList<ResultDetail> = mutableListOf()
+        val warnings: MutableList<ResultDetail> = mutableListOf()
+
+        db.transact { txn ->
```

Comment:
Change this transaction so it only covers the first two fetches, then use another for the create, that way you are not having all this other non-database code inside the transaction block.  Transactions are only really useful for writes, so you do not gain anything from having all three calls in one transaction.

Reply:
We can't use separate transactions here due to the way we are locking tasks - the fetch is running .forUpdate() so that the row is locked until the transaction commits, preventing another instance of the function from grabbing the same task.